File: src/322.cpp
#include <iostream>
#include <vector>
#include <cmath>

class Solution {
public:
    int coinChange(std::vector<int> & coins, int amount) {
        if (amount == 0) {
            return 0;
        }

        std::vector<int> dp(amount + 1);
        dp[0] = 0;
        for (int i = 1; i <= amount; i ++) {
            dp[i] = amount + 1;
        }

        for (int i = 0; i < coins.size(); i++) {
//            printf("%d", coins[i]);
            if (coins[i] < amount) {
                dp[coins[i]] = 1;
            }
        }



        for (int i = 1; i < amount + 1; i++) {
            for (int j = 0; j < coins.size(); j++) {
                if (i >= coins[j]) {
                    int temp = 0;
                    temp = dp[i - coins[j]] + 1;
                    dp[i] = min(dp[i], temp);
                }
            }
        }

        if (dp[amount] == amount + 1) {
            return -1;
        }
        return dp[amount];
    }

    int min(int & a, int & b) {
        if (a >= b) {
            return b;
        } else {
            return a;
        }

    }
};


int main() {
//    std::cout << "Hello, World!" << std::endl;
    int amount = 1;
    std::vector <int> coins(1);
    coins = {1};
    Solution s;
    int result;
    result = s.coinChange(coins, amount);
    printf("%d", result);


}

File: src/322.py
import sys
##github

class Solution:
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        if amount == 0:
            return 0

        dp = [sys.maxsize for i in range(amount + 1)]
        dp[0] = 0
        for item in coins:
            if item <= amount:
                dp[item] = 1

        for i in range(1, amount + 1):
            for item in coins:
                if i >= item:
                    dp[i] = min(dp[i], dp[i - item] + 1)
        if dp[amount] == sys.maxsize:
            return -1
        return dp[amount]


###
###这里cpp中 max_int + 1 = min_int  所以不能使用max_int 赋值
###
